test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/RuntimeTypeTestsTest.scala

```diff
@@ -0,0 +1,539 @@
+
+    testArrayArrayObject(false, new Array[Array[Int]](5))
+
+    testArrayArrayObject(false, new Array[Object](5))
+    testArrayArrayObject(false, new Array[Seq[_]](5))
+    testArrayArrayObject(false, new Array[List[_]](5))
+    testArrayArrayObject(false, new Array[String](5))
+    testArrayArrayObject(false, new Array[Serializable](5))
+    testArrayArrayObject(false, new Array[Cloneable](5))
+
+    testArrayArrayObject(false, new Array[Int](5))
+    testArrayArrayObject(false, new Array[Long](5))
+
+    testArrayArrayObject(false, List(5))
+    testArrayArrayObject(false, new Object)
+
+    @inline def testArrayArraySeq(expected: Boolean, value: Any): Unit = {
+      test(expected, value, classOf[Array[Array[Seq[_]]]],
+          _.isInstanceOf[Array[Array[Seq[_]]]], _.asInstanceOf[Array[Array[Seq[_]]]])
+    }
+
+    testArrayArraySeq(true, new Array[Array[Seq[_]]](5))
+    testArrayArraySeq(true, new Array[Array[List[_]]](5))
+
+    testArrayArraySeq(false, new Array[Array[Object]](5))
+    testArrayArraySeq(false, new Array[Seq[_]](5))
+    testArrayArraySeq(false, new Array[Array[Array[Seq[_]]]](5))
+  }
+
+  @Test def nothingType(): Unit = {
+    // scalac refuses to compile x.isInstanceOf[Nothing], so we fake it
+    val isInstance: Any => Boolean = _ => false
+
+    // Force Scala 3 to accept our code
+    implicit val ct: ClassTag[Nothing] = ClassTag(classOf[Nothing])
+
+    @inline def testNothing(expected: Boolean, value: Any): Unit =
+      test[Nothing](expected, value, classOf[Nothing], isInstance, _.asInstanceOf[Nothing])
+
+    testNothing(false, "a")
+    testNothing(false, List(5))
+
+    /* The behavior of casting for null is not consistent, due to how scalac
+     * desugars things. This is true both on JS and on the JVM. To test null,
+     * we therefore only test the isInstance tests.
+     */
+    assertFalse(classOf[Nothing].isInstance(null))
+  }
+
+  @Test def nullType(): Unit = {
+    // scalac refuses to compile x.isInstanceOf[Null], so we fake it
+    val isInstance: Any => Boolean = _ => false
+
+    // Force Scala 3 to accept our code
+    implicit val ct: ClassTag[Null] = ClassTag(classOf[Null])
+
+    testNullValue(classOf[Null], isInstance, _.asInstanceOf[Null])
+
+    @inline def testNull(expected: Boolean, value: Any): Unit =
+      test(expected, value, classOf[Null], isInstance, _.asInstanceOf[Null])
+
+    testNull(false, "a")
+    testNull(false, 0)
+    testNull(false, false)
+    testNull(false, ())
+    testNull(false, List(5))
+  }
+
+}
+
+object RuntimeTypeTestsTest {
+
+  @noinline
+  def assertAlmostSame(msg: String, expected: Any, actual: Any): Unit = {
+    /* "Almost" same is like "same", except that we tolerate `equals` for
+     * boxed classes, since the instances could be unboxed and reboxed by the
+     * monomorphic code in the lambdas.
+     */
+
+    expected match {
+      case _:Boolean | _:Char | _:Byte | _:Short | _:Int | _:Long | _:Float | _:Double =>
```

Comment:
It's different from checking `isPrimitive()` because, if we get `expected.getClass()`, it's going to be the *boxed* class, which is not a primitive. So that doesn't help.

I have added a big comment to explain why it's fine.

Reply:
I meant different than `targetClass.isPrimitive()`. For example: We use `assertAlmostSame` behind `if (!targetClass.isPrimitive())`. I do not understand when this is necessary (i.e. when is `targetClass.isPrimitive()` true, but the `assertEquals` branch is required).

Or another way of looking at this: If whether or not a thing we are testing (i.e. a `targetClass` we pass in) is primitive is statically known at call-site of `test`, we might want to avoid determining it dynamically.

Now, if it is not the case that `targetClass.isPrimitive()` iff when we need `assertEquals`, then this might result in  too many flags / arguments, so some dynamic determination might be better overall.

Reply:
@sjrd did you see this comment? Otherwise LGTM.

Reply:
Ah no I had missed that.

It's not the same as `targetClass.isPrimitive()` either. The target class could be one of the boxed classes, or any of their super classes or super interfaces, and we could still have an expected value that is a primitive, requiring `assertEquals`. We're not testing a property of the target class here. It really is a property of the values we're comparing.